### PR TITLE
[Test] More resilience for token invalidation test

### LINKED ISSUE #67347: testExpiredTokensDeletedAfterExpiration fails with "Error invalidating refresh_token"
**Build scan**: https://gradle-enterprise.elastic.co/s/m2opl5u64rlmc/console-log?task=:x-pack:plugin:security:internalClusterTest

**Repro line**: `./gradlew ':x-pack:plugin:security:internalClusterTest' --tests "org.elasticsearch.xpack.security.authc.TokenAuthIntegTests.testExpiredTokensDeletedAfterExpiration" -Dtests.seed=CAFFBB05C6B0FD31 -Dtests.security.manager=true -Dtests.locale=en-ZA -Dtests.timezone=America/Recife -Druntime.java=11`

**Reproduces locally?**: No

**Applicable branches**: `master`

**Failure history**: [Failing occasionally](https://build-stats.elastic.co/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:now-30d,mode:quick,to:now))&_a=(columns:!(_source),index:b646ed00-7efc-11e8-bf69-63c8ef516157,interval:auto,query:(language:lucene,query:testExpiredTokensDeletedAfterExpiration),sort:!(process.time-start,desc))) after a flurry of failures back in December.

**Failure excerpt**:

```

org.elasticsearch.xpack.security.authc.TokenAuthIntegTests > testExpiredTokensDeletedAfterExpiration FAILED |  
-- | --
  | java.lang.AssertionError: |  
  | Expected: an empty collection |  
  | but: <[org.elasticsearch.ElasticsearchException: Elasticsearch exception [type=exception, reason=Error invalidating refresh_token: ]]> |  
  | at __randomizedtesting.SeedInfo.seed([CAFFBB05C6B0FD31:CBDEAF1CFDF6B220]:0) |  
  | at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18) |  
  | at org.junit.Assert.assertThat(Assert.java:956) |  
  | at org.junit.Assert.assertThat(Assert.java:923) |  
  | at org.elasticsearch.xpack.security.authc.TokenAuthIntegTests.testExpiredTokensDeletedAfterExpiration(TokenAuthIntegTests.java:208)

```

### PR DESCRIPTION
**Pull request by ywangd**
This PR adds more resilience to testExpiredTokensDeletedAfterExpiration
by wrapping the empty error assertion into an assertBusy block. In rare
cases, it is possible for search to return a deleted but not yet
refreshed document. Trying to invalidate this document then result into
MissDocumentException. This state is transient and should go away once
all shards refresh.

Resolves: #67347

### CODE REVIEW

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java [hunk 1]
```diff
@@ -197,22 +197,23 @@ public void testExpiredTokensDeletedAfterExpiration() throws Exception {
...
         assertThat(invalidateRefreshTokenResponse.getErrors(), empty());
 
-        // 99% of the time, this will already be zero, but if not ensure it goes to zero within the allowed timeframe
-        if (invalidateRefreshTokenResponse.getPreviouslyInvalidatedTokens() > 0) {
+        // 99% of the time, this will already be empty, but if not ensure it goes to empty within the allowed timeframe
+        if (false == invalidateRefreshTokenResponse.getErrors().isEmpty()) {
             assertBusy(() -> {
                 var newResponse = restClient.security().invalidateToken(invalidateRefreshTokenRequest, SECURITY_REQUEST_OPTIONS);
-                assertThat(newResponse.getPreviouslyInvalidatedTokens(), equalTo(0));
+                assertThat(newResponse.getErrors(), empty());
```
Comment (ywangd) [inline x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java:216]: I was the reviewer for the previous change (#64757). But somehow I didn't realise that we should wrap `getErrors()` here instead of `getPreviouslyInvalidatedTokens()` because:
1. The original failure (#56903) was about `getErrors()`.
2. `getPreviouslyInvalidatedTokens()` should always return `0` since the refersh token is never invalidated before. If the cluster behaves really weirdly, we can probably expect `getInvalidatedTokens()` to be `1`, but not `getPreviouslyInvalidatedTokens()`.

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java [hunk 2]
```diff
@@ -197,22 +197,22 @@ public void testExpiredTokensDeletedAfterExpiration() throws Exception {
 
         // Weird testing behaviour ahead...
         // invalidating by access token (above) is a Get, but invalidating by refresh token (below) is a Search
-        // In a multi node cluster, in a small % of cases, the search might find a document that has been invalidated but not yet deleted
+        // In a multi node cluster, in a small % of cases, the search might find a document that has been deleted but not yet refreshed
         // from that node's shard.
```
Comment (tvernum) [inline x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java:201]:
```suggestion
        // in that node's shard.
```

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java [hunk 3]
```diff
@@ -197,22 +197,22 @@ public void testExpiredTokensDeletedAfterExpiration() throws Exception {
...
-        // Our assertion, therefore, is that an attempt to invalidate the (already invalidated) refresh token must not actually invalidate
+        // Our assertion, therefore, is that an attempt to invalidate the refresh token must not actually invalidate
         // anything (concurrency controls must prevent that), nor may return any errors,
-        // but it might _temporarily_ find an "already invalidated" token.
+        // but it might _temporarily_ find an "already deleted" token.
         final InvalidateTokenRequest invalidateRefreshTokenRequest = InvalidateTokenRequest.refreshToken(refreshToken);
         InvalidateTokenResponse invalidateRefreshTokenResponse = restClient.security().invalidateToken(
             invalidateRefreshTokenRequest, SECURITY_REQUEST_OPTIONS);
         assertThat(invalidateRefreshTokenResponse.getInvalidatedTokens(), equalTo(0));
-        assertThat(invalidateRefreshTokenResponse.getErrors(), empty());
```
Comment (tvernum) [inline x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java:209]: One of the problems here is that we have this very complex test, but nothing testing the more normal cases.
As far as I can tell there is nothing to test what happens if you try to invalidate an access token and then invalidate its refresh token. What's that behaviour supposed to be?

And this test is trying to test a behaviour that we mostly don't care about - specifically what happens if you try to invalidate a refresh token after it's been deleted from the index? The answer is it should either behave like a token that never existed, or it should behave like a token that was recently invalidated. We don't really mind which.

I don't think we should ever return errors though - it's not the caller's problem that the cleanup job just ran.

Comment (ywangd) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java:209]: > but nothing testing the more normal cases

We can (and should) add more tests. But I'd prefer a separate PR. This one is just to tame the number of CI failures.

> And this test is trying to test a behaviour that we mostly don't care about - specifically what happens if you try to invalidate a refresh token after it's been deleted from the index?

If we go by the test name literally, to me, `testExpiredTokensDeletedAfterExpiration` means it is sufficient to ensure that the expired token document is deleted once the cleanup job runs. It has nothing to do with subsuquent invalidations after the deletion. And like you said, any subsequent invalidation should just behave like the document never exists and we do have a separate test (`testInvalidateNotValidRefreshTokens`) for that. If this is the case, I'd say the test is sufficient after the [second assertBusy block](https://github.com/elastic/elasticsearch/blob/7.14/x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java#L198-L216) which asserts the document no longer exists after triggering the cleanup job.

So a simpler fix is to just drop all the code after it. We can then have separate PRs to add tests for more common use cases, e.g. invalidate refresh token after access token is invalidated etc.

Comment (tvernum) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java:209]: Part of my concern here is that this test says it's OK to return errors for a period of time after the deletion occurs, as long as they eventually go away.

Why is that? That seems like it's just accepting that there is a bug that we don't want to fix.
And due to the weirdness in these tests overall, it's hard to tell whether that's a bug that manifests more generally, or whether it's just an edge case that we don't care about and should stop testing for.

Because it sounds like we don't care, and we're just trying to get the test to run.  
Why do these assertions model the behaviour that we actually want from the code?

Comment (ywangd) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java:209]: I added [a new test](https://github.com/elastic/elasticsearch/pull/75319/commits/41d596eb806f499c81c54a313f04648a06e99723) for "deleting access and refresh tokens one after the other with randomness" to cover the more realistic use case as you suggested.

As for this test, the elaborated code indeed suggests for a bug. There are also other bugs in TokenService and they are all more or less caused by the same dirty read problem. All of them are quite rare and this one is even less a practical issue. So we didn't get time around to tackle them, plus they all have some complexities. Even for this one, code wise it is likely easy to skip any `DocumentMissingException` and just returns normally. But I am not sure whether this would be covering up other subtle and likely very rare problems. In my opinion, it is at least safer to report a false error instead of silently skip the error (unless we are perfectly sure). All these sounds like a bit too much for a PR that is meant to reduce our CI backlog. I can raise an issue to track the bug, but I'd prefer to keep it separate from this PR.

**Review by tvernum — APPROVED** [review]
LGTM.
We need to fix this test. We clearly have some underlying issues around requests that come in at the same time as the cleanup job, but they're more theoretical than of practical concern.

**Comment by ywangd** [comment]
> We need to fix this test. We clearly have some underlying issues around requests that come in at the same time as the cleanup job, but they're more theoretical than of practical concern.

I agree there are more issues to fix. But I think it is more about the code itself rather than the test. The test is a reflect on how the code logic being non-ideal. Once we get around to fix the code, the test should be also be fixed as part of it.
